### PR TITLE
Add authentication flow to the Observability-Portal

### PR DESCRIPTION
**Pull request by Shairam**

### CODE REVIEW

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/TrustAllX509TrustManager.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package io.cellery.observability.api;
```
Comment (sinthuja) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/TrustAllX509TrustManager.java]: Missing license header

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
+package io.cellery.observability.api.internal;
```
Comment (sinthuja) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Missing license header

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    private static final Logger log = Logger.getLogger(AggregatedRequestsAPI.class);
+
+    public static JSONObject getClientCredentials() throws IOException {
+        BufferedReader bufReader = null;
+        JSONObject obj2 = null;
+        try {
+
+            HttpClient client = getAllSSLClient();
+
+            HttpPost request =
```
Comment (sinthuja) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Please declare these strings as constants.

File: docker/sp/pom.xml [hunk 4]
```diff
@@ -157,6 +157,26 @@
             <groupId>io.cellery.observability</groupId>
             <artifactId>io.cellery.observability.k8s.api.server.client</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.oltu.oauth2</groupId>
+            <artifactId>org.apache.oltu.oauth2.client</artifactId>
+            <version>1.0.0</version>
```
Comment (sinthuja) [inline docker/sp/pom.xml]: Include the dependency versions to the root pom.xml, and the infer the dependency in the child modules.

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.TrustManager;
+import javax.ws.rs.GET;
+import javax.ws.rs.OPTIONS;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
```
Comment (nadundesilva) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java]: Please remove unwanted additional lines

File: components/global/core/io.cellery.observability.api/pom.xml [hunk 6]
```diff
@@ -56,6 +56,37 @@
             <groupId>org.wso2.carbon.datasources</groupId>
             <artifactId>org.wso2.carbon.datasource.core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
+            <version>4.3.6</version>
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/pom.xml]: Let's more versions to parent pom with proper properties.

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        this.clientId = id;
+        this.clientSecret = secret;
+    }
+
+    @GET
+    @Path("/requestToken/{authCode}")
+    @Produces("application/json")
+    public Response getTokens(@PathParam("authCode") String authCode) throws APIInvocationException {
+        try {
+           cancelCheck();
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java]: Formatting

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+            SSLContext sc = SSLContext.getInstance("TLS");
+            sc.init(null, new TrustManager[]{new TrustAllX509TrustManager()}, new java.security.SecureRandom());
+            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
+            HttpsURLConnection.setDefaultHostnameVerifier(new HostnameVerifier() {
+                public boolean verify(String string, SSLSession ssls) {
+                    return true;
+                }
+            });
+
+        } catch (NoSuchAlgorithmException | KeyManagementException e) {
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java]: Let's add a meaningful error message, than "Error"

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            while ((line = bufReader.readLine()) != null) {
+                builder.append(line);
+                builder.append(System.lineSeparator());
+            }
+
+            obj2 = new JSONObject(builder.toString());
+            return obj2;
+
+        } catch (KeyStoreException | NoSuchAlgorithmException | KeyManagementException e) {
+            log.error("Message", e);
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Always add a meaningful error message. For an example "Error while retrieving client credentials. "

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            StringBuilder builder = new StringBuilder();
+
+            String line;
+
+            while ((line = bufReader.readLine()) != null) {
+                builder.append(line);
+                builder.append(System.lineSeparator());
+            }
+
+            obj2 = new JSONObject(builder.toString());
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Let's use meaningful variable names..

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+            obj.put("owner", "admin");
+            obj.put("grantType", "authorization_code");
+            obj.put("saasApp", " true");
+            request.setHeader("Authorization", "Basic YWRtaW46YWRtaW4=");
+            request.setHeader("content-type", "application/json");
+            request.setEntity(new StringEntity(obj.toString()));
+
+            HttpResponse response = client.execute(request);
+
+            bufReader = new BufferedReader(new InputStreamReader(
```
Comment (hasinthaindrajee) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Don't we need to close this input stream ?

File: components/global/portal/io.cellery.observability.ui/src/components/SignIn.js [hunk 12]
```diff
@@ -60,9 +54,22 @@ const styles = (theme) => ({
...
+        right: 0,
+        bottom: 0,
+        left: 0,
+        width: "200px",
+        height: "100px"
+
     }
 });
 
+const idpAddress = "wso2-apim";
```
Comment (nadundesilva) [inline components/global/portal/io.cellery.observability.ui/src/components/SignIn.js]: Let's move this to constants.js file for now and add a "TODO" comment. (Need to move to portal config later)

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+public class RegisterClient {
+
+    private static final Logger log = Logger.getLogger(AggregatedRequestsAPI.class);
+
+    public static JSONObject getClientCredentials() throws IOException {
+        BufferedReader bufReader = null;
+        InputStreamReader inputStreamReader = null;
+        JSONObject jsonObject = null;
+        try {
+
```
Comment (sinthuja) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/internal/RegisterClient.java]: Remove unnecessary new lines.

File: components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @GET
+    @Path("/requestToken/{authCode}")
+    @Produces("application/json")
+    public Response getTokens(@PathParam("authCode") String authCode) throws APIInvocationException {
+        try {
+            cancelCheck();
+            OAuthClientRequest request = OAuthClientRequest
+                    .tokenLocation("https://gateway.cellery-system:9443/oauth2/token?")
```
Comment (sinthuja) [inline components/global/core/io.cellery.observability.api/src/main/java/io/cellery/observability/api/UserAuthenticationAPI.java]: Use the constants defined above without using the inline string.

File: docker/sp/pom.xml (hunk 4 above)
Comment (sinthuja) [inline docker/sp/pom.xml]: You can remove the versions from this pom.xml given that it's defined in the root pom.xml

**Pull request merged**